File: bench/memory/misalign.zig
const std = @import("std");

const N = 1_000_000;

fn printResult(name: []const u8, total_ns: u64, iters: u64) !void {
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {d:>8} ns {d:>12}\n", .{ name, total_ns / iters, iters });
}

pub fn main() !void {
    const alloc = std.heap.page_allocator;
    const w = std.io.getStdOut().writer();
    try w.print("{s:<40} {s:>12} {s:>12}\n", .{ "Benchmark", "Time", "Iterations" });
    try w.print("{s}\n", .{"-" ** 66});

    // Aligned access
    {
        const arr = try alloc.alloc(i64, N);
        defer alloc.free(arr);
        @memset(arr, 42);

        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i64 = 0;
            for (arr) |x| sum +%= x;
            std.mem.doNotOptimizeAway(sum);
            total += timer.read();
        }
        try printResult("BM_AlignedAccess", total, iters);
    }

    // Misaligned access
    {
        const buf = try alloc.alloc(u8, N * @sizeOf(i64) + 3);
        defer alloc.free(buf);
        const ptr: [*]align(1) i64 = @ptrCast(buf.ptr + 3);
        for (0..N) |i| ptr[i] = 42;

        var total: u64 = 0;
        var iters: u64 = 0;
        while (total < 1_000_000_000) : (iters += 1) {
            var timer = try std.time.Timer.start();
            var sum: i64 = 0;
            for (0..N) |i| sum +%= ptr[i];
            std.mem.doNotOptimizeAway(sum);
            total += timer.read();
        }
        try printResult("BM_MisalignedAccess", total, iters);
    }
}

File: bench/memory/misalign.cpp
#include <benchmark/benchmark.h>

#include <cstring>
#include <vector>

static void BM_AlignedAccess(benchmark::State& state) {
  std::vector<int64_t> arr(1'000'000, 42);

  for (auto _ : state) {
    int64_t sum = 0;
    for (auto i = 0u; i < arr.size(); ++i) {
      sum += arr[i];
    }
    benchmark::DoNotOptimize(sum);
  }
}

static void BM_MisalignedAccess(benchmark::State& state) {
  constexpr int N = 1'000'000;
  std::vector<char> buf(N * sizeof(int64_t) + 3);
  char* p = buf.data() + 3;

  int64_t val = 42;
  for (int i = 0; i < N; ++i) {
    memcpy(p + i * sizeof(int64_t), &val, sizeof(int64_t));
  }

  for (auto _ : state) {
    int64_t sum = 0;
    for (int i = 0; i < N; ++i) {
      int64_t v;
      memcpy(&v, p + i * sizeof(int64_t), sizeof(int64_t));
      sum += v;
    }
    benchmark::DoNotOptimize(sum);
  }
}

BENCHMARK(BM_AlignedAccess);
BENCHMARK(BM_MisalignedAccess);

BENCHMARK_MAIN();
